### PR TITLE
Break on failure to deploy cert

### LINKED ISSUE #4002: Incorrect success condition in nginx
`Cannot find a cert or key directive in /etc/nginx/sites-enabled/site.conf for set(['sitecom']). VirtualHost was not modified.`

After getting this message, we still ask about Easy or Secure, and then print a success message. That is incorrect.

### PR DESCRIPTION
**Pull request by ohemorange**
Fixes #4002.

### CODE REVIEW

**Comment by ohemorange** [comment]
Is this what we want?

```
(venv) root@ubuntu-512mb-sfo2-letsencrypt:/home/erica/certbot# certbot run --cert-name newname --nginx -vv
[...]
Cert not yet due for renewal

You have an existing certificate that has exactly the same domains or certificate name you requested and isn't close to expiry.
(ref: /etc/letsencrypt/renewal/newname.conf)

What would you like to do?
-------------------------------------------------------------------------------
1: Attempt to reinstall this existing certificate
2: Renew & replace the cert (limit ~5 per 7 days)
-------------------------------------------------------------------------------
Select the appropriate number [1-2] then [enter] (press 'c' to cancel): 1
Keeping the existing certificate
Problem in /etc/nginx/sites-enabled/functorkitten.conf: expected directive for ssl_certificate in the Nginx config but did not find it.
Encountered exception:
Traceback (most recent call last):
  File "/home/erica/certbot/certbot/client.py", line 385, in deploy_certificate
    fullchain_path=fullchain_path)
  File "/home/erica/certbot/certbot-nginx/certbot_nginx/configurator.py", line 196, in deploy_cert
    "VirtualHost was not modified.", vhost.filep, vhost.names)
PluginError: ('Cannot find a cert or key directive in %s for %s. VirtualHost was not modified.', '/etc/nginx/sites-enabled/functorkitten.conf', set(['nineteen.functorkitten.xyz']))

Calling registered functions
Exiting abnormally:
Traceback (most recent call last):
  File "/home/erica/certbot/venv/bin/certbot", line 11, in <module>
    load_entry_point('certbot', 'console_scripts', 'certbot')()
  File "/home/erica/certbot/certbot/main.py", line 849, in main
    return config.func(config, plugins)
  File "/home/erica/certbot/certbot/main.py", line 579, in run
    lineage.chain, lineage.fullchain)
  File "/home/erica/certbot/certbot/client.py", line 385, in deploy_certificate
    fullchain_path=fullchain_path)
  File "/home/erica/certbot/certbot-nginx/certbot_nginx/configurator.py", line 196, in deploy_cert
    "VirtualHost was not modified.", vhost.filep, vhost.names)
PluginError: ('Cannot find a cert or key directive in %s for %s. VirtualHost was not modified.', '/etc/nginx/sites-enabled/functorkitten.conf', set(['nineteen.functorkitten.xyz']))


(venv) root@ubuntu-512mb-sfo2-letsencrypt:/home/erica/certbot# certbot run --cert-name newname --nginx 
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Starting new HTTPS connection (1): acme-v01.api.letsencrypt.org
Cert not yet due for renewal

You have an existing certificate that has exactly the same domains or certificate name you requested and isn't close to expiry.
(ref: /etc/letsencrypt/renewal/newname.conf)

What would you like to do?
-------------------------------------------------------------------------------
1: Attempt to reinstall this existing certificate
2: Renew & replace the cert (limit ~5 per 7 days)
-------------------------------------------------------------------------------
Select the appropriate number [1-2] then [enter] (press 'c' to cancel): 1
Keeping the existing certificate 
```

Note `-vv`

**Comment by ohemorange** [comment]
@ above nvm, probably. it was an improperly formed raise. why it wasn't printing, no idea.

**Comment by pde** [comment]
What does this error look like to the user now (without `-v`)?

**Comment by ohemorange** [comment]
```
(venv) root@ubuntu# certbot run --nginx --cert-name domain.com
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Starting new HTTPS connection (1): acme-v01.api.letsencrypt.org
Cert not yet due for renewal

You have an existing certificate that has exactly the same domains or certificate name you requested and isn't close to expiry.
(ref: /etc/letsencrypt/renewal/domain.com.conf)

What would you like to do?
-------------------------------------------------------------------------------
1: Attempt to reinstall this existing certificate
2: Renew & replace the cert (limit ~5 per 7 days)
-------------------------------------------------------------------------------
Select the appropriate number [1-2] then [enter] (press 'c' to cancel): 1
Keeping the existing certificate
Cannot find a cert or key directive in /etc/nginx/sites-enabled/domain.com.conf for set(['domain.com']). VirtualHost was not modified.
```

**Comment by pde** [comment]
And what does it look like if you pick option 2, `renew`?

**Comment by ohemorange** [comment]
(Aside: because we haven't fixed the reading-from-config thing, it didn't pick up my previous `--staging` flag and crashed lol. But we've decided to kick that for this release.)

```
(venv) root@ubuntu# certbot run --nginx --cert-name domain.com --staging
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Starting new HTTPS connection (1): acme-staging.api.letsencrypt.org
Cert not yet due for renewal

You have an existing certificate that has exactly the same domains or certificate name you requested and isn't close to expiry.
(ref: /etc/letsencrypt/renewal/domain.com.conf)

What would you like to do?
-------------------------------------------------------------------------------
1: Attempt to reinstall this existing certificate
2: Renew & replace the cert (limit ~5 per 7 days)
-------------------------------------------------------------------------------
Select the appropriate number [1-2] then [enter] (press 'c' to cancel): 2
Renewing an existing certificate
Performing the following challenges:
tls-sni-01 challenge for domain.com
/home/user/certbot/venv/local/lib/python2.7/site-packages/OpenSSL/rand.py:58: UserWarning: implicit cast from 'char *' to a different pointer type: will be forbidden in the future (check that the types are as you expect; use an explicit ffi.cast() if they are correct)
  result_code = _lib.RAND_bytes(result_buffer, num_bytes)
Waiting for verification...
Cleaning up challenges
Generating key (2048 bits): /etc/letsencrypt/keys/0069_key-certbot.pem
Creating CSR: /etc/letsencrypt/csr/0084_csr-certbot.pem
Cannot find a cert or key directive in /etc/nginx/sites-enabled/domain.com.conf for set(['domain.com']). VirtualHost was not modified.

IMPORTANT NOTES:
 - Congratulations! Your certificate and chain have been saved at
   /etc/letsencrypt/live/domain.com/fullchain.pem. Your cert will expire
   on 2017-04-10. To obtain a new or tweaked version of this
   certificate in the future, simply run certbot again with the
   "certonly" option. To non-interactively renew *all* of your
   certificates, run "certbot renew"
```

**Comment by pde** [comment]
It feels like we might want to have an little bit more explanation to the user of what happened here. They got a cert, but it wasn't installed!

One way to do this might be to wrap the call to `le_client.deploy_cert` in `main.run` with a `try: except:` that in the `action in ("renew", "newcert")` cases logs `Certbot was able to obtain a certificate for you, but wasn't able to install it; you'll need to do that by hand.`.

It also feels like the word "Congratulations!" is in appropriate in that case, but that's definitely another ticket.

**Comment by ohemorange** [comment]
@pde, why aren't we printing the error message in `client.py:deploy_certificate`? Do we only print that on failure to restart, not failure to deploy? Not immediately clear what all these `ErrorHandler`s are doing.

**Comment by pde** [comment]
We could print it (`try: except: ... raise`) in `client.deploy_certificate` too, yes. It should be appropriately mindful that it might be called from `main.install` too, though that subcommand is kind of janky and broken still.

**Comment by bmw** [comment]
FWIW, taking the relevant bits of my half written comment here, see #3303 if you really want to know more about our `ErrorHandler` and how it can go wrong.

**Comment by ohemorange** [comment]
```
(venv) root@ubuntut# certbot --cert-name domain.com --nginx
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Cert not yet due for renewal

You have an existing certificate that has exactly the same domains or certificate name you requested and isn't close to expiry.
(ref: /etc/letsencrypt/renewal/domain.com.conf)

What would you like to do?
-------------------------------------------------------------------------------
1: Attempt to reinstall this existing certificate
2: Renew & replace the cert (limit ~5 per 7 days)
-------------------------------------------------------------------------------
Select the appropriate number [1-2] then [enter] (press 'c' to cancel): 1
Keeping the existing certificate
Cannot find a cert or key directive in /etc/nginx/sites-enabled/domain.com.conf for set(['domain.com']). VirtualHost was not modified.

IMPORTANT NOTES:
 - Unable to install the certificate
(venv) root@ubuntu# certbot --cert-name domain.com --nginx
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Cert not yet due for renewal

You have an existing certificate that has exactly the same domains or certificate name you requested and isn't close to expiry.
(ref: /etc/letsencrypt/renewal/domain.com.conf)

What would you like to do?
-------------------------------------------------------------------------------
1: Attempt to reinstall this existing certificate
2: Renew & replace the cert (limit ~5 per 7 days)
-------------------------------------------------------------------------------
Select the appropriate number [1-2] then [enter] (press 'c' to cancel): 2
Renewing an existing certificate
Performing the following challenges:
tls-sni-01 challenge for domain.com
Waiting for verification...
Cleaning up challenges
Generating key (2048 bits): /etc/letsencrypt/keys/0071_key-certbot.pem
Creating CSR: /etc/letsencrypt/csr/0086_csr-certbot.pem
Cannot find a cert or key directive in /etc/nginx/sites-enabled/domain.com.conf for set(['domain.com']). VirtualHost was not modified.

IMPORTANT NOTES:
 - Unable to install the certificate
 - Congratulations! Your certificate and chain have been saved at
   /etc/letsencrypt/live/domain.com/fullchain.pem. Your cert will expire
   on 2017-04-10. To obtain a new or tweaked version of this
   certificate in the future, simply run certbot again with the
   "certonly" option. To non-interactively renew *all* of your
   certificates, run "certbot renew"
```

**Comment by bmw** [comment]
LGTM. Push changes when you're ready and I'll review.

File: certbot-nginx/certbot_nginx/configurator.py [hunk 1]
```diff
@@ -191,11 +191,9 @@ def deploy_cert(self, domain, cert_path, key_path,
                         vhost.filep, vhost.names)
         except errors.MisconfigurationError as error:
             logger.debug(error)
-            logger.warning(
-                "Cannot find a cert or key directive in %s for %s. "
-                "VirtualHost was not modified.", vhost.filep, vhost.names)
             # Presumably break here so that the virtualhost is not modified
-            return False
+            raise errors.PluginError("Cannot find a cert or key directive in {0} for {1}. "
```
Comment (bmw) [inline certbot-nginx/certbot_nginx/configurator.py]: There's no unit test coverage here. I think this is valuable, but I'm fine delaying this as long as one one of us test these lines manually.

**Comment by bmw** [comment]
Client tests should also be changed (they need to mock out the reporter), however, I may have found a larger problem in this area (unrelated to this PR). 0.10.0 will come out someday...

File: certbot-nginx/certbot_nginx/configurator.py (hunk 1 above)
Comment (bmw) [inline certbot-nginx/certbot_nginx/configurator.py]: I tested this manually and it works as expected. Up to you if you want to add tests before merging.

**Comment by bmw** [comment]
Looking at https://travis-ci.org/certbot/certbot/jobs/190810100#L737 (because Coveralls isn't done), it doesn't look like the test hit the right lines.

**Pull request merged**